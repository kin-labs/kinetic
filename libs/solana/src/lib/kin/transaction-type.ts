export enum TransactionType {
  Unknown = -1,
  None = 0,
  Earn = 1,
  Spend = 2,
  P2P = 3,
}
