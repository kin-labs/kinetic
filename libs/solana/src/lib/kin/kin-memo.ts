import { MAGIC_BYTE, MAX_TRANSACTION_TYPE, MAX_VERSION } from './kin-memo-constants'
import { TransactionType } from './transaction-type'

// Memo implements the Agora memo specification.
//
// Spec: https://github.com/kinecosystem/agora-api
export class KinMemo {
  buffer: Buffer
  constructor(buf: Buffer) {
    this.buffer = buf
  }

  static from(b: Buffer): KinMemo {
    const buf = Buffer.alloc(b.length)
    b.copy(buf)
    return new this(buf)
  }

  static fromB64String(s: string, strict = true): KinMemo | undefined {
    const raw = Buffer.from(s, 'base64')
    const m = KinMemo.from(raw)
    if (!KinMemo.isValid(m, strict)) {
      throw new Error('invalid memo')
    }

    return m
  }

  static new(version: number, type: TransactionType, index: number, fk: Buffer): KinMemo {
    if (fk.length > 29) {
      throw new Error('invalid foreign key length')
    }
    if (version > 7) {
      throw new Error('invalid version')
    }
    if (type < 0) {
      throw new Error('cannot use unknown transaction type')
    }

    const b = Buffer.alloc(32)

    // encode magic byte + version
    b[0] = MAGIC_BYTE
    b[0] |= version << 2

    // encode transaction type
    b[0] |= (type & 0x7) << 5
    b[1] = (type & 0x18) >> 3

    // encode index
    b[1] |= (index & 0x3f) << 2
    b[2] = (index & 0x3fc0) >> 6
    b[3] = (index & 0xc000) >> 14

    if (fk.byteLength > 0) {
      b[3] |= (fk[0] & 0x3f) << 2
      // insert the rest of the fk. since each loop references fk[n] and fk[n+1], the upper bound is offset by 3 instead of 4.
      for (let i = 4; i < 3 + fk.byteLength; i++) {
        // apply last 2-bits of current byte
        // apply first 6-bits of next byte
        b[i] = (fk[i - 4] >> 6) & 0x3
        b[i] |= (fk[i - 3] & 0x3f) << 2
      }

      // if the foreign key is less than 29 bytes, the last 2 bits of the FK can be included in the memo
      if (fk.byteLength < 29) {
        b[fk.byteLength + 3] = (fk[fk.byteLength - 1] >> 6) & 0x3
      }
    }

    return new this(b)
  }

  static isValid(m: KinMemo, strict?: boolean): boolean {
    if (Number(m.buffer[0] & 0x3) != MAGIC_BYTE) {
      return false
    }

    if (m.transactionTypeRaw() == -1) {
      return false
    }

    if (!strict) {
      return true
    }

    if (m.version() > MAX_VERSION) {
      return false
    }

    return m.transactionType() >= 0 && m.transactionType() <= MAX_TRANSACTION_TYPE
  }

  // version returns the memo encoding version.
  version(): number {
    return (this.buffer[0] & 0x1c) >> 2
  }

  // transactionType returns the type of the transaction the memo is
  // attached to.
  transactionType(): TransactionType {
    const raw = this.transactionTypeRaw()
    if (raw >= 0 && raw <= MAX_TRANSACTION_TYPE) {
      return raw
    }

    return TransactionType.Unknown
  }

  // transactionTypeRaw returns the type of the transaction the memo is
  // attached to, even if it is unsupported by this implementation. It should
  // only be used as a fallback if the raw value is needed when transactionType()
  // yields TransactionType.Unknown.
  transactionTypeRaw(): TransactionType {
    return (this.buffer[0] >> 5) | ((this.buffer[1] & 0x3) << 3)
  }

  // index returns the index of the app the transaction relates to.
  index(): number {
    const a = Number(this.buffer[1]) >> 2
    const b = Number(this.buffer[2]) << 6
    const c = Number(this.buffer[3] & 0x3) << 14
    return a | b | c
  }

  // foreignKey returns an identifier in an auxiliary service that contains
  // additional information related to the transaction.
  foreignKey(): Buffer {
    const fk = Buffer.alloc(29)

    for (let i = 0; i < 28; i++) {
      fk[i] |= this.buffer[i + 3] >> 2
      fk[i] |= (this.buffer[i + 4] & 0x3) << 6
    }

    // We only have 230 bits, which results in
    // our last fk byte only having 6 'valid' bits
    fk[28] = this.buffer[31] >> 2

    return fk
  }
}
