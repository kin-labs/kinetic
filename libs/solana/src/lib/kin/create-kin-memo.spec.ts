import { createKinMemo } from './create-kin.memo'
import { KinMemo } from './kin-memo'

import { TransactionType } from './transaction-type'

const index = 1
describe('createKinMemo', () => {
  it('create memo, type None', () => {
    const memo = createKinMemo({ index })
    const parsed = KinMemo.fromB64String(memo)

    expect(memo).toEqual('BQQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=')
    expect(parsed.transactionType()).toEqual(TransactionType.None)
    expect(parsed.index()).toEqual(1)
  })

  it('create memo, type Spend', () => {
    const memo = createKinMemo({ index, type: TransactionType.Spend })
    const parsed = KinMemo.fromB64String(memo)

    expect(memo).toEqual('RQQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=')
    expect(parsed.transactionType()).toEqual(TransactionType.Spend)
    expect(parsed.index()).toEqual(1)
  })

  it('create memo, type Earn', () => {
    const memo = createKinMemo({ index, type: TransactionType.Earn })
    const parsed = KinMemo.fromB64String(memo)

    expect(memo).toEqual('JQQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=')
    expect(parsed.transactionType()).toEqual(TransactionType.Earn)
    expect(parsed.index()).toEqual(1)
  })

  it('create memo, type P2P', () => {
    const memo = createKinMemo({ index, type: TransactionType.P2P })
    const parsed = KinMemo.fromB64String(memo)

    expect(memo).toEqual('ZQQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=')
    expect(parsed.transactionType()).toEqual(TransactionType.P2P)
    expect(parsed.index()).toEqual(1)
  })

  it('create memo with reference data', () => {
    const reference = '1234567890abcdefghijklmnopqrstuvwxyz'
    const memo = createKinMemo({ index, reference })
    const parsed = KinMemo.fromB64String(memo)

    expect(memo).toMatchSnapshot()
    expect(parsed.transactionType()).toEqual(TransactionType.None)
    expect(parsed.foreignKey().toString('base64')).toEqual(`${reference}AAA=`)
    expect(parsed.index()).toEqual(1)
  })

  it('fail creating a memo with reference data that is too long', () => {
    const reference = `1234567890abcdefghijklmnopqrstuvwxyz1234`

    try {
      createKinMemo({ index, reference })
    } catch (e) {
      expect(e.message).toEqual('invalid foreign key length')
    }
  })
})
