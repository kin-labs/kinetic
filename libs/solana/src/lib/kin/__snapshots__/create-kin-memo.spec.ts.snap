// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createKinMemo create memo with reference data 1`] = `"BQQAXLfhn7vy3xttxl1/CmKMSmadimquym6/DnPMAgA="`;
